config/sources/common.conf

```diff
@@ -6,19 +6,180 @@
+	# Attention: having KERNELSOURCE set here usually means it is a fork, vendor kernel, or other non-strict mainline kernel.
+	if [[ -n ${KERNELSOURCE} ]]; then
+		if [[ -z ${KERNEL_MAJOR_MINOR} ]]; then
+			display_alert "KERNELSOURCE is set, but KERNEL_MAJOR_MINOR is unset" "please make sure they match" "warn"
+			((problems++))
+		fi
+		if [[ -z ${KERNELBRANCH} ]]; then
+			display_alert "KERNELSOURCE is set, but KERNELBRANCH is unset" "please make sure they match" "warn"
+			((problems++))
+		fi
+	fi
+
+	# KERNEL_MAJOR_MINOR is the key to all this stuff. It must be set.
+	if [[ -z ${KERNEL_MAJOR_MINOR} ]]; then
+		display_alert "KERNEL_MAJOR_MINOR is unset" "common_defaults_for_mainline" "warn"
+		((problems++))
+	fi
+
+	# exit with error if some breaking problem detected above.
+	if ((problems > 0)); then
+		exit_with_error "Please fix the above kernel configuration problems (check warnings above)."
+	fi
+
+	# Small debuggings to help when everything else fails and all sanity is gone.
+	if [[ -z ${KERNELSOURCE} ]]; then
+		display_alert "KERNELSOURCE is unset" "common_defaults_for_mainline" "debug"
+	else
+		display_alert "KERNELSOURCE is set by board/family/hooks to '${KERNELSOURCE}'" "common_defaults_for_mainline" "debug"
+	fi
+
+	if [[ -z ${KERNELBRANCH} ]]; then
+		display_alert "KERNELBRANCH is unset" "common_defaults_for_mainline" "debug"
+	else
+		display_alert "KERNELBRANCH is set by board/family/hooks to '${KERNELBRANCH}'" "common_defaults_for_mainline" "debug"
+	fi
+
+	if [[ -z ${LINUXFAMILY} ]]; then
+		display_alert "LINUXFAMILY is unset" "common_defaults_for_mainline" "debug"
+	else
+		display_alert "LINUXFAMILY is set by board/family/hooks to '${LINUXFAMILY}'" "common_defaults_for_mainline" "debug"
+	fi
+
+	if [[ -z ${KERNELPATCHDIR} ]]; then
+		display_alert "KERNELPATCHDIR is unset" "common_defaults_for_mainline" "debug"
+	else
+		display_alert "KERNELPATCHDIR is set by board/family/hooks to '${KERNELPATCHDIR}'" "common_defaults_for_mainline" "debug"
+	fi
+
+	if [[ -z ${KERNEL_PATCH_ARCHIVE_BASE} ]]; then # @TODO: this is a bad idea, remove
+		display_alert "KERNEL_PATCH_ARCHIVE_BASE is unset" "common_defaults_for_mainline" "debug"
+	else
+		display_alert "KERNEL_PATCH_ARCHIVE_BASE is set by board/family/hooks to '${KERNELPATCHDIR}'" "common_defaults_for_mainline" "debug"
+	fi
+
+	if [[ -z ${LINUXCONFIG} ]]; then
+		display_alert "LINUXCONFIG is unset" "common_defaults_for_mainline" "debug"
+	else
+		display_alert "LINUXCONFIG is set by board/family/hooks to '${LINUXCONFIG}'" "common_defaults_for_mainline" "debug"
+	fi
+
+	# Generic defaultings, used for all kernels (including vendor/legacy/3rd-party and mainline).
+	# If KERNEL_PATCH_ARCHIVE_BASE is unset, set it to LINUXFAMILY.
+	if [[ -z ${KERNEL_PATCH_ARCHIVE_BASE} ]]; then # @TODO: this is a bad idea, remove
+		display_alert "KERNEL_PATCH_ARCHIVE_BASE is unset; using LINUXFAMILY: '${LINUXFAMILY}'" "common_defaults_for_mainline" "debug"
```

Comment:
Do we really need another variable? I mean if someone would need to override KERNELPATCHDIR, they can do that directly, why throw another variable in the mix?

Reply:
True, and indeed, many families already do set KERNELPATCHDIR (eg sunxi). I'd prefer that LINUXFAMILY was used consistently, unfortunately, there are many instances where the patch directory does not match LINUXFAMILY, so I introduced this as a way to override/migrate slowly over time.

Reply:
Instead we will end up having 3 different ways to override the same thing and people will be using all three ways in one family or another.

Reply:
We are entirely too tolerant of such things, note the other methods as deprecated, and break them when the time comes.  If no one is fixing their junk then their junk stops working.  🤷‍♂️